compiler/ic/store.nim

```diff
@@ -0,0 +1,149 @@
+
+template config(m: Module): ConfigRef = m.ast.sh.config
+
+proc rodFile*(conf: ConfigRef; m: PSym): AbsoluteFile =
+  ## find the target rodfile of a given module
+  result = AbsoluteFile toFullPath(conf, m.info.fileIndex)
+  result = result.changeFileExt "rod"
+
+proc writeModuleInto(m: Module; fn: AbsoluteFile; value = hash(m)) =
+  ## write the module into the given rodfile path; the hash value
+  ## allows us to save an extra computation of the hash
+  let noSerializeSubstitute = m.ast.sh.config
+  var stream = newFileStream($fn, fmWrite)
+  m.ast.sh.config = nil
+  try:
+    # the following three values constitute the rodfile's "header"
+    freeze(version, stream)
+    freeze(hash noSerializeSubstitute, stream)
+    freeze(value, stream)
+    # try to reduce some buffer churn on the read
+    freeze(compress(freeze m), stream)
+  finally:
+    m.ast.sh.config = noSerializeSubstitute
+    close stream
+
+proc readHeader(stream: Stream): Header {.raises: [].} =
+  ## query a rod file stream to see if it is worth attempting to use
+  try:
+    setPosition(stream, 0)
+    thaw(stream, result.version)
+    if result.version != version:
+      result.msg = "version `$1` doesn't match `$2`" % [ $result.version,
+                                                         $version ]
+    else:
+      thaw(stream, result.config)
+      thaw(stream, result.hash)
+      result.ok = true
+  except CatchableError as e:
+    result.msg = e.msg
+
+proc readHeader(fn: AbsoluteFile): Header =
+  ## query a rod file to see if it is worth attempting to use
+  if not fileExists fn:
+    result.msg = $fn & ": file not found"
+  else:
+    try:
+      var stream = newFileStream($fn, fmRead)
+      try:
+        result = readHeader stream
+      finally:
+        close stream
+    except CatchableError as e:
+      result.msg = e.msg
+
+proc composeFilename(config: ConfigRef; name: string): AbsoluteFile =
+  ## turn an arbitrary name into a .rod filename
+  let dir = getNimcacheDir(config) / RelativeDir "rod"
+  createDir dir
+  result = dir / addFileExt(RelativeFile mangle(name), "rod")
+
+template handleFileError(config: ConfigRef; fn: AbsoluteFile; e: typed) =
+  ## handle some kind of error thrown up by a rod file
+  when defined(release):
+    removeFile fn
+  else:
+    internalError(config, $fn & ": " & e.msg)
+
+proc tryReadModule*(config: ConfigRef; fn: AbsoluteFile): Option[Module] =
+  ## populated option if reading the module was successful
+  if not fileExists fn: return none(Module)
+  var stream = newFileStream($fn, fmRead)
+  try:
+    let reply = readHeader stream
+    if reply.ok:
+      if hash(config) != reply.config:
+        # NOTE: we just bail if the configs don't hash equivalently
+        return
+      # try to reduce some buffer churn
+      var snap = newStringOfCap(getFileSize $fn)
+      thaw(stream, snap)
+      var m = thaw[Module] uncompress(snap)
+      if hash(m.ast) != reply.hash:
```

Comment:
Sorry, I meant here.

Reply:
I think it's more important that the `hash()` for a `Table` key supplies our concept of equivalency.  Can it be done with `MD5`?  Sure.  Should it?  Why?

One reason to leave it this way is that `frosty` has *debug* code to audit de/serialization via an included `Hash` value.

Reply:
First off, sha1/md5 are widely used for this purpose (together with SHA256, Sha512, etc).

Why not to use `std/hashes`? Because it may break, hash has overloads for pointer types that just cast them to int (this is done ofc for good reasons). Adding fields in serialized objects may break the hash functions unexpectedly. It's just not it's use-case! 

Moreover you need to be clear on what exactly are you testing. I suppose that is, the saved file is being retrieved without corruption. By adding the check at later stage, you effectively don't trust your serialization lib and test it's output.

My thought was that since serializing to a string dumps every type into a flat, sequence of bytes,
it's clever to use this in order to produce your hash of the contents. By using hash you effectively add another indirection that needs to be maintained and tested.

Reply:
I see you use hash in frosty a lot. Maybe you should consider other ways for testing, like making roundtrips, exact equality `==` operators, writing to a string and testing its length, using `setPosition` and reading from there, etc. If you need a performant lib that's also well tested, I might know smth.

Reply:
What we're testing is that the current compiler's idea of what's important about the data in the rodfile matches the prior compiler's idea.  If we can't agree on the value of what we've read, then we know that we need to assume that the contents of the file are stale.  The valuations will be used across the code to measure equivalency between packed objects, so it's not the data that's important -- it's what it *means*.

Reply:
> I see you use hash in frosty a lot. Maybe you should consider other ways for testing, like making roundtrips, exact equality `==` operators, writing to a string and testing its length, using `setPosition` and reading certain parts of the stream, etc. If you need a performant lib that's also well tested, I might know smth. ;)


By all means, feel free to recommend something.  Better yet, submit a PR to `frosty` that adds a benchmark comparison to the repository so we can measure the improvement given identical input.  There are existing tests that you can use to ensure your alternative meets the requirements of PackedAst.

`frosty` hasn't been optimized for static data as `jason` has -- I think my approach there sucks and I'm waiting until concepts work better to try again with `frosty` -- so there is clearly a lot of room for improvement.

Reply:
Man, hash is NOT equivalency, its used to shortcircuit `==` eveen in [tables](https://github.com/nim-lang/Nim/blob/5514b299ebc1de4ef0c1c182d4f21a02e3c2fa90/lib/pure/collections/hashcommon.nim#L60) Lets not make another lib war, do whats correct.

Reply:
What's confusing about this?